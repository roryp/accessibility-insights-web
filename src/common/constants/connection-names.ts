// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
export class ConnectionNames {
    public static readonly devTools: string = 'insights-dev-tools';
    public static readonly debugToolsTelemetry: string = 'insights-debug-tools-telemetry';
}
