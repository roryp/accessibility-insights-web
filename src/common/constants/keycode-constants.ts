// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
export class KeyCodeConstants {
    public static readonly TAB: number = 9;
    public static readonly ENTER: number = 13;
    public static readonly ALT: number = 18;
    public static readonly SPACEBAR: number = 32;
}
