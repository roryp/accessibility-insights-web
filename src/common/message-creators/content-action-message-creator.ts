// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { BaseActionPayload } from 'background/actions/action-payloads';
import { ContentPayload } from 'background/actions/content-actions';
import { ActionMessageDispatcher } from 'common/message-creators/types/dispatcher';
import * as React from 'react';

import { ActionInitiators } from '../action/action-initiator';
import {
    CONTENT_HYPERLINK_OPENED,
    CONTENT_PAGE_OPENED,
    TelemetryEventSource,
} from '../extension-telemetry-events';
import { Messages } from '../messages';
import { TelemetryDataFactory } from '../telemetry-data-factory';

export class ContentActionMessageCreator {
    constructor(
        private readonly telemetryFactory: TelemetryDataFactory,
        private readonly source: TelemetryEventSource,
        private readonly dispatcher: ActionMessageDispatcher,
    ) {}

    public openContentPage = (
        event: React.MouseEvent<any> | MouseEvent,
        contentPath: string,
    ): void => {
        const messageType = Messages.Telemetry.Send;
        const telemetry = this.telemetryFactory.withTriggeredByAndSource(event, this.source);
        const payload = {
            eventName: CONTENT_PAGE_OPENED,
            telemetry: { ...telemetry, contentPath },
        };

        this.dispatcher.dispatchMessage({
            messageType,
            payload,
        });
    };

    public openContentHyperLink = (
        event: React.MouseEvent<any> | MouseEvent,
        href: string,
    ): void => {
        const messageType = Messages.Telemetry.Send;
        const telemetry = this.telemetryFactory.withTriggeredByAndSource(event, this.source);
        const payload = {
            eventName: CONTENT_HYPERLINK_OPENED,
            telemetry: { ...telemetry, href },
        };
        this.dispatcher.dispatchMessage({
            messageType: messageType,
            payload,
        });
    };

    public openContentPanel = (
        event: React.MouseEvent<HTMLElement> | MouseEvent,
        contentPath: string,
        contentTitle: string,
    ): void => {
        const messageType = Messages.ContentPanel.OpenPanel;
        const telemetry = this.telemetryFactory.withTriggeredByAndSource(event, this.source);
        const payload: ContentPayload = {
            telemetry,
            contentPath,
            contentTitle,
        };

        this.dispatcher.dispatchMessage({
            messageType,
            payload,
        });
    };

    public closeContentPanel = (): void => {
        const messageType = Messages.ContentPanel.ClosePanel;
        const telemetry = this.telemetryFactory.fromDetailsViewNoTriggeredBy();
        const payload: BaseActionPayload = {
            telemetry,
        };

        this.dispatcher.dispatchMessage({
            messageType: messageType,
            payload,
        });
    };

    private openExternalLink = (
        event: React.MouseEvent<any> | MouseEvent,
        details: { href: string },
    ): void => {
        this.openContentHyperLink(event, details.href);
    };

    public initiators: Pick<ActionInitiators, 'openExternalLink'> = {
        openExternalLink: this.openExternalLink,
    };
}
