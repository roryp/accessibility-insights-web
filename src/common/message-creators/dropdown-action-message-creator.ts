// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { BaseActionPayload } from 'background/actions/action-payloads';
import { ActionMessageDispatcher } from 'common/message-creators/types/dispatcher';

import { TelemetryEventSource } from '../extension-telemetry-events';
import { Message } from '../message';
import { Messages } from '../messages';
import { TelemetryDataFactory } from '../telemetry-data-factory';

export class DropdownActionMessageCreator {
    constructor(
        private readonly telemetryFactory: TelemetryDataFactory,
        private readonly dispatcher: ActionMessageDispatcher,
    ) {}

    public openPreviewFeaturesPanel(
        event: React.MouseEvent<HTMLElement>,
        source: TelemetryEventSource,
    ): void {
        const messageType = Messages.PreviewFeatures.OpenPanel;
        const telemetry = this.telemetryFactory.withTriggeredByAndSource(event, source);
        const payload: BaseActionPayload = {
            telemetry,
        };
        const message: Message = {
            messageType: messageType,
            payload,
        };
        this.dispatcher.dispatchMessage(message);
    }

    public openScopingPanel(
        event: React.MouseEvent<HTMLElement>,
        source: TelemetryEventSource,
    ): void {
        const messageType = Messages.Scoping.OpenPanel;
        const telemetry = this.telemetryFactory.withTriggeredByAndSource(event, source);
        const payload: BaseActionPayload = {
            telemetry,
        };
        const message = {
            messageType: messageType,
            payload,
        };
        this.dispatcher.dispatchMessage(message);
    }

    public openSettingsPanel(
        event: React.MouseEvent<HTMLElement>,
        source: TelemetryEventSource,
    ): void {
        const messageType = Messages.SettingsPanel.OpenPanel;
        const telemetry = this.telemetryFactory.forSettingsPanelOpen(event, source, 'menu');
        const payload: BaseActionPayload = {
            telemetry,
        };
        const message = {
            messageType: messageType,
            payload,
        };
        this.dispatcher.dispatchMessage(message);
    }

    public openDebugTools(): void {
        this.dispatcher.dispatchType(Messages.DebugTools.Open);
    }
}
