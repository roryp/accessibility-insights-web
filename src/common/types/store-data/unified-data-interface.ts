// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { BoundingRectangle } from 'electron/platform/android/android-scan-results';
import { GuidanceLink } from '../../../scanner/rule-to-links-mappings';
import { ScanIncompleteWarningId } from '../scan-incomplete-warnings';

// this is similar to `TestEngine` interface from axe-core
export interface ScanEngineProperties {
    name: string;
    version: string;
}

export interface ApplicationProperties {
    name: string;
    version: string;
    environmentName?: string;
}

export interface OSProperties {
    name: string;
    version?: string;
    userAgent?: string;
}

export interface ViewPortProperties {
    height?: number;
    width?: number;
    dpi?: number;
}

export interface PlatformData {
    osInfo: OSProperties;
    viewPortInfo: ViewPortProperties;
    deviceName?: string;
}

export interface TargetAppData {
    name: string;
    url?: string;
}

export interface ToolData {
    scanEngineProperties: ScanEngineProperties;
    applicationProperties: ApplicationProperties;
}

export type ScanMetadata = {
    timestamp: string;
    toolData: ToolData;
    targetAppInfo: TargetAppData;
    deviceName?: string;
};

export interface UnifiedRule {
    id: string;
    description: string;
    url: string;
    guidance: GuidanceLink[];
}

export interface UnifiedScanResultStoreData {
    results: UnifiedResult[];
    rules: UnifiedRule[];
    platformInfo?: PlatformData;
    toolInfo?: ToolData;
    targetAppInfo?: TargetAppData;
    timestamp?: string;
    scanIncompleteWarnings?: ScanIncompleteWarningId[];
    screenshotData?: ScreenshotData;
}

export interface InstancePropertyBag {
    [property: string]: any;
}

export type StoredInstancePropertyBag = InstancePropertyBag;

export type UnifiedIdentifiers = {
    identifier: string;
    conciseName: string;
} & InstancePropertyBag;

export type UnifiedDescriptors = {
    snippet?: string;
    boundingRectangle?: BoundingRectangle;
} & InstancePropertyBag;

export type UnifiedResolution = {
    howToFixSummary: string;
} & InstancePropertyBag;

export type FormattableResolution = {
    howToFix: string;
    formatAsCode?: string[];
};

export type UnifiedFormattableResolution = {
    howToFixFormat?: FormattableResolution;
} & UnifiedResolution;

export interface UnifiedResult {
    uid: string;
    status: InstanceResultStatus;
    ruleId: string;
    identifiers: UnifiedIdentifiers;
    descriptors: UnifiedDescriptors;
    resolution: UnifiedResolution;
}

export type InstanceResultStatus = 'pass' | 'fail' | 'unknown';

export interface ScreenshotData {
    base64PngData: string;
}
