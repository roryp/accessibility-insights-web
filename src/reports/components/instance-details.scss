// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

@import '../../common/styles/colors.scss';
@import '../../common/styles/fonts.scss';

@mixin fill-available-width {
    width: -moz-available;
    /* WebKit-based browsers will ignore this. */
    width: -webkit-fill-available;
    /* Mozilla-based browsers will ignore this. */
    width: fill-available;
}

.instance-details-card {
    border-radius: 4px;
    border: 1px solid $card-border;
    outline-style: 'border-style';
    box-shadow: 0px 0.6px 1.8px $box-shadow-108, 0px 3.2px 7.2px $box-shadow-132;
    margin-bottom: 16px;
    cursor: pointer;
    @include fill-available-width;
}

.instance-details-card-container.selected {
    outline: 5px solid $communication-tint-10;
}

.instance-details-card.selected {
    border: 1px solid transparent;
}

.instance-details-card:focus {
    outline: 2px solid $primary-text;
    outline-offset: 2px;
}

.instance-details-card.selected:focus {
    outline-offset: 8px;
}

.instance-details-card:hover {
    box-shadow: 0px 8px 10px $box-shadow-108, 0px 8px 10px $box-shadow-132;
}

.report-instance-table {
    background: $neutral-0;

    display: table;
    table-layout: fixed;
    @include fill-available-width;

    border-radius: inherit;

    th {
        padding: 12px 20px;

        vertical-align: top;
    }

    .row {
        top: calc(50% - 20px / 2);
        border-bottom: 0.5px solid $neutral-10;
    }

    .row:last-child {
        border-bottom: none;
    }
}

.report-instance-table .label {
    font-size: 14px;
    line-height: 20px;
    font-family: $semiBoldFontFamily;
    color: $primary-text;
    text-align: left;
    width: 70px;
}

.report-instance-table .instance-list-row-content {
    color: $secondary-text;
    padding: 12px 20px;
    font-size: 14px;
    line-height: 20px;
    align-items: flex-end;
    display: flex;
}

.report-instance-table .instance-list-row-content.content-snipppet {
    font-family: Consolas, Segoe UI, Courier New, monospace;
    white-space: normal;
    word-wrap: break-word;
    word-break: break-word;
}

table.report-instance-table {
    border-collapse: collapse;
    overflow-x: auto;
    word-break: break-word;
}

@media screen and (max-width: 480px) {
    .report-instance-table tr {
        display: block;
    }

    .report-instance-table td {
        display: block;

        .label {
            text-align: left;
        }

        .instance-list-row-content {
            text-align: right;
        }
    }

    .report-instance-table td::before {
        float: left;
    }

    .report-instance-table .label {
        padding: 14px 20px 6px 20px;
    }

    .report-instance-table .instance-list-row-content {
        padding: 6px 20px 14px 20px;
    }
}
