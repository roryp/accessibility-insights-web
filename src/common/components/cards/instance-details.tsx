// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as classNames from 'classnames';
import { KeyCodeConstants } from 'common/constants/keycode-constants';
import { NamedFC } from 'common/react/named-fc';
import { CardResult } from 'common/types/store-data/card-view-model';
import { forOwn, isEmpty } from 'lodash';
import * as React from 'react';
import {
    instanceDetailsCard,
    instanceDetailsCardContainer,
    reportInstanceTable,
    selected,
} from 'reports/components/instance-details.scss';

import {
    CardRowDeps,
    PropertyConfiguration,
} from '../../../common/configs/unified-result-property-configurations';
import { CardSelectionMessageCreator } from '../../../common/message-creators/card-selection-message-creator';
import {
    StoredInstancePropertyBag,
    TargetAppData,
    UnifiedRule,
} from '../../../common/types/store-data/unified-data-interface';
import { UserConfigurationStoreData } from '../../types/store-data/user-configuration-store';
import { InstanceDetailsFooter, InstanceDetailsFooterDeps } from './instance-details-footer';

export const instanceCardAutomationId = 'instance-card';

export type InstanceDetailsDeps = {
    getPropertyConfigById: (id: string) => PropertyConfiguration;
    cardSelectionMessageCreator: CardSelectionMessageCreator;
} & CardRowDeps &
    InstanceDetailsFooterDeps;

export type InstanceDetailsProps = {
    deps: InstanceDetailsDeps;
    result: CardResult;
    index: number;
    userConfigurationStoreData: UserConfigurationStoreData;
    targetAppInfo: TargetAppData;
    rule: UnifiedRule;
};

export const InstanceDetails = NamedFC<InstanceDetailsProps>('InstanceDetails', props => {
    const { result, index, deps, userConfigurationStoreData, rule, targetAppInfo } = props;

    const isHighlightSupported: boolean = deps.cardInteractionSupport.supportsHighlighting;

    const renderCardRowsForPropertyBag = (propertyBag: StoredInstancePropertyBag) => {
        let propertyIndex = 0;
        const cardRows = [];
        forOwn(propertyBag, (propertyData, propertyName) => {
            const propertyConfig = deps.getPropertyConfigById(propertyName);
            if (!isEmpty(propertyConfig)) {
                const CardRow = propertyConfig.cardRow;
                ++propertyIndex;
                cardRows.push(
                    <CardRow
                        deps={deps}
                        propertyData={propertyData}
                        index={index}
                        key={`${propertyName}-${propertyIndex}`}
                    />,
                );
            }
        });
        return <>{cardRows}</>;
    };

    const cardClickHandler = (event: React.SyntheticEvent): void => {
        if (isHighlightSupported) {
            deps.cardSelectionMessageCreator.toggleCardSelection(result.ruleId, result.uid, event);
        }
    };

    const cardKeyPressHandler = (event: React.KeyboardEvent<any>): void => {
        if (
            event.keyCode === KeyCodeConstants.ENTER ||
            event.keyCode === KeyCodeConstants.SPACEBAR
        ) {
            event.preventDefault();
            cardClickHandler(event);
        }
    };

    const instanceDetailsCardStyling = classNames({
        [instanceDetailsCard]: true,
        [selected]: isHighlightSupported ? result.isSelected : false,
    });

    const instanceDetailsCardContainerStyling = classNames({
        [instanceDetailsCardContainer]: true,
        [selected]: isHighlightSupported ? result.isSelected : false,
    });

    const cardAriaLabel = `${
        result.identifiers && result.identifiers.identifier ? result.identifiers.identifier : ''
    } card`;

    return (
        <div
            data-automation-id={instanceCardAutomationId}
            className={instanceDetailsCardContainerStyling}
            role="table"
        >
            <div
                className={instanceDetailsCardStyling}
                tabIndex={0}
                onClick={cardClickHandler}
                onKeyDown={cardKeyPressHandler}
                aria-selected={result.isSelected}
                aria-label={cardAriaLabel}
                role="row"
            >
                <div role="gridcell">
                    <table className={reportInstanceTable}>
                        <tbody>
                            {renderCardRowsForPropertyBag(result.identifiers)}
                            {renderCardRowsForPropertyBag(result.descriptors)}
                            {renderCardRowsForPropertyBag(result.resolution)}
                        </tbody>
                    </table>
                    <InstanceDetailsFooter
                        deps={deps}
                        result={result}
                        userConfigurationStoreData={userConfigurationStoreData}
                        rule={rule}
                        targetAppInfo={targetAppInfo}
                    />
                </div>
            </div>
        </div>
    );
});
