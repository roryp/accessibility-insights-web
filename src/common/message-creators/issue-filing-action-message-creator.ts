// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { BaseActionPayload, FileIssuePayload } from 'background/actions/action-payloads';
import { ActionMessageDispatcher } from 'common/message-creators/types/dispatcher';

import { ToolData } from 'common/types/store-data/unified-data-interface';
import { FILE_ISSUE_CLICK, TelemetryEventSource } from '../extension-telemetry-events';
import { Message } from '../message';
import { Messages } from '../messages';
import { TelemetryDataFactory } from '../telemetry-data-factory';
import { CreateIssueDetailsTextData } from '../types/create-issue-details-text-data';

type SupportedMouseEvent = React.MouseEvent<HTMLElement> | React.SyntheticEvent<Element, Event>;

export class IssueFilingActionMessageCreator {
    constructor(
        private readonly dispatcher: ActionMessageDispatcher,
        private readonly telemetryFactory: TelemetryDataFactory,
        private readonly source: TelemetryEventSource,
    ) {}

    public openSettingsPanel(event: React.MouseEvent<HTMLElement>): void {
        const messageType = Messages.SettingsPanel.OpenPanel;
        const telemetry = this.telemetryFactory.forSettingsPanelOpen(
            event,
            this.source,
            'fileIssueSettingsPrompt',
        );
        const payload: BaseActionPayload = {
            telemetry,
        };
        this.dispatcher.dispatchMessage({
            messageType: messageType,
            payload,
        });
    }

    public trackFileIssueClick(event: React.MouseEvent<HTMLElement>, service: string): void {
        const telemetry = this.telemetryFactory.forFileIssueClick(event, this.source, service);
        this.dispatcher.sendTelemetry(FILE_ISSUE_CLICK, telemetry);
    }

    public fileIssue(
        event: SupportedMouseEvent,
        serviceKey: string,
        issueData: CreateIssueDetailsTextData,
        toolData: ToolData,
    ): void {
        const messageType = Messages.IssueFiling.FileIssue;
        const telemetry = this.telemetryFactory.forFileIssueClick(event, this.source, serviceKey);
        const payload: FileIssuePayload = {
            telemetry,
            issueData,
            service: serviceKey,
            toolData,
        };
        const message: Message = {
            messageType,
            payload,
        };
        this.dispatcher.dispatchMessage(message);
    }
}
