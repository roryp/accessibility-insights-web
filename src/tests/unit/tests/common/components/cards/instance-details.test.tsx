// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import {
    allCardInteractionsSupported,
    noCardInteractionsSupported,
} from 'common/components/cards/card-interaction-support';
import {
    InstanceDetails,
    InstanceDetailsDeps,
    InstanceDetailsProps,
} from 'common/components/cards/instance-details';
import {
    AllPropertyTypes,
    CardRowProps,
    PropertyConfiguration,
} from 'common/configs/unified-result-property-configurations';
import { KeyCodeConstants } from 'common/constants/keycode-constants';
import { CardSelectionMessageCreator } from 'common/message-creators/card-selection-message-creator';
import { NamedFC, ReactFCWithDisplayName } from 'common/react/named-fc';
import { UnifiedResolution, UnifiedResult } from 'common/types/store-data/unified-data-interface';
import { shallow } from 'enzyme';
import * as React from 'react';
import { instanceDetailsCard } from 'reports/components/instance-details.scss';
import { IMock, It, Mock, Times } from 'typemoq';

import { exampleUnifiedResult } from './sample-view-model-data';

describe('InstanceDetails', () => {
    let props: InstanceDetailsProps;
    let deps: InstanceDetailsDeps;
    let getPropertyConfigByIdMock: IMock<(id: string) => PropertyConfiguration>;
    let cardSelectionMessageCreatorMock: IMock<CardSelectionMessageCreator>;
    let resultStub: UnifiedResult;
    let indexStub: number;

    beforeEach(() => {
        getPropertyConfigByIdMock = Mock.ofInstance(_ => null);
        cardSelectionMessageCreatorMock = Mock.ofType(CardSelectionMessageCreator);
        resultStub = exampleUnifiedResult;
        indexStub = 22;

        deps = {
            getPropertyConfigById: getPropertyConfigByIdMock.object,
            cardSelectionMessageCreator: cardSelectionMessageCreatorMock.object,
            cardInteractionSupport: allCardInteractionsSupported,
        } as InstanceDetailsDeps;
        props = {
            deps,
            result: resultStub,
            index: indexStub,
        } as InstanceDetailsProps;
    });

    it('renders', () => {
        setupGetPropertyConfigByIdMock();
        cardSelectionMessageCreatorMock
            .setup(mock => mock.toggleCardSelection(It.isAnyString(), It.isAnyString(), It.isAny()))
            .verifiable(Times.never());

        const testSubject = shallow(<InstanceDetails {...props} />);

        expect(testSubject.getElement()).toMatchSnapshot();
        cardSelectionMessageCreatorMock.verifyAll();
    });

    it('dispatches the card selection message when card is clicked', () => {
        setupGetPropertyConfigByIdMock();
        const eventStub = {} as React.SyntheticEvent;

        cardSelectionMessageCreatorMock
            .setup(mock => mock.toggleCardSelection(It.isAnyString(), It.isAnyString(), eventStub))
            .verifiable(Times.once());

        const wrapper = shallow(<InstanceDetails {...props} />);
        const divElem = wrapper.find(`.${instanceDetailsCard}`);
        expect(divElem.length).toBe(1);

        divElem.simulate('click', eventStub);

        cardSelectionMessageCreatorMock.verifyAll();
    });

    it('does not dispatch the card selection message when card is clicked if highlighting is not supported', () => {
        deps = {
            ...deps,
            cardInteractionSupport: noCardInteractionsSupported,
        };
        props = {
            ...props,
            deps,
        };

        setupGetPropertyConfigByIdMock();

        cardSelectionMessageCreatorMock
            .setup(mock => mock.toggleCardSelection(It.isAnyString(), It.isAnyString(), It.isAny()))
            .verifiable(Times.never());

        const wrapper = shallow(<InstanceDetails {...props} />);
        const divElem = wrapper.find(`.${instanceDetailsCard}`);
        expect(divElem.length).toBe(1);

        divElem.simulate('click');

        cardSelectionMessageCreatorMock.verifyAll();
    });

    const supportedKeyCodes = [KeyCodeConstants.ENTER, KeyCodeConstants.SPACEBAR];
    it.each(supportedKeyCodes)(
        'dispatches the card selection message when key with keycode %s is pressed',
        keyCode => {
            const preventDefaultMock = jest.fn();
            const eventStub = { keyCode: keyCode, preventDefault: preventDefaultMock } as any;
            setupGetPropertyConfigByIdMock();

            cardSelectionMessageCreatorMock
                .setup(mock =>
                    mock.toggleCardSelection(It.isAnyString(), It.isAnyString(), eventStub),
                )
                .verifiable(Times.once());

            const wrapper = shallow(<InstanceDetails {...props} />);
            const divElem = wrapper.find(`.${instanceDetailsCard}`);
            expect(divElem.length).toBe(1);

            divElem.simulate('keydown', eventStub);

            cardSelectionMessageCreatorMock.verifyAll();
            expect(preventDefaultMock).toHaveBeenCalled();
        },
    );

    it('renders nothing when there is no card row configuration for the property / no property', () => {
        props.result.identifiers = {
            identifier: 'test-id',
            conciseName: 'test-concise-name',
            'this-property-does-not-have-config': 'some value',
        };
        props.result.descriptors = {};
        props.result.resolution = {} as UnifiedResolution;

        setupGetPropertyConfigByIdMock();

        const testSubject = shallow(<InstanceDetails {...props} />);
        expect(testSubject.getElement()).toMatchSnapshot();
    });

    it.each([true, false])('isSelected drives the styling for the card', isSelected => {
        props.result.isSelected = isSelected;
        setupGetPropertyConfigByIdMock();

        const testSubject = shallow(<InstanceDetails {...props} />);
        expect(testSubject.getElement()).toMatchSnapshot();
    });

    function getCardRowStub(name: string): ReactFCWithDisplayName<CardRowProps> {
        return NamedFC<CardRowProps>(name, _ => null);
    }

    function setupGetPropertyConfigByIdMock(): void {
        AllPropertyTypes.forEach(propertyType => {
            const propertyConfigurationStub: PropertyConfiguration = {
                cardRow: getCardRowStub(propertyType),
            };
            getPropertyConfigByIdMock
                .setup(mock => mock(propertyType))
                .returns(() => propertyConfigurationStub);
        });
    }
});
