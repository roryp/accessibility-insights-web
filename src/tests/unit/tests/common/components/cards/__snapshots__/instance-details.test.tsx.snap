// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstanceDetails isSelected drives the styling for the card 1`] = `
<div
  className="instanceDetailsCardContainer selected"
  data-automation-id="instance-card"
  role="table"
>
  <div
    aria-label="test-id card"
    aria-selected={true}
    className="instanceDetailsCard selected"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="row"
    tabIndex={0}
  >
    <div
      role="gridcell"
    >
      <table
        className="reportInstanceTable"
      >
        <tbody>
          <React.Fragment />
          <React.Fragment />
          <React.Fragment />
        </tbody>
      </table>
      <InstanceDetailsFooter
        deps={
          Object {
            "cardInteractionSupport": Object {
              "supportsCopyFailureDetails": true,
              "supportsHighlighting": true,
              "supportsIssueFiling": true,
            },
            "cardSelectionMessageCreator": proxy {
              "___id": "BCDF5CE5-F0DF-40B7-8BA0-69DF395033C8",
              "collapseAllRules": [Function],
              "dispatcher": undefined,
              "expandAllRules": [Function],
              "source": undefined,
              "telemetryFactory": undefined,
              "toggleCardSelection": [Function],
              "toggleRuleExpandCollapse": [Function],
              "toggleVisualHelper": [Function],
            },
            "getPropertyConfigById": [Function],
          }
        }
        result={
          Object {
            "descriptors": Object {},
            "identifiers": Object {
              "conciseName": "test-concise-name",
              "identifier": "test-id",
              "this-property-does-not-have-config": "some value",
            },
            "isSelected": true,
            "resolution": Object {},
            "ruleId": "image-alt",
            "status": "fail",
            "uid": "some-guid-here",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`InstanceDetails isSelected drives the styling for the card 2`] = `
<div
  className="instanceDetailsCardContainer"
  data-automation-id="instance-card"
  role="table"
>
  <div
    aria-label="test-id card"
    aria-selected={false}
    className="instanceDetailsCard"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="row"
    tabIndex={0}
  >
    <div
      role="gridcell"
    >
      <table
        className="reportInstanceTable"
      >
        <tbody>
          <React.Fragment />
          <React.Fragment />
          <React.Fragment />
        </tbody>
      </table>
      <InstanceDetailsFooter
        deps={
          Object {
            "cardInteractionSupport": Object {
              "supportsCopyFailureDetails": true,
              "supportsHighlighting": true,
              "supportsIssueFiling": true,
            },
            "cardSelectionMessageCreator": proxy {
              "___id": "BCDF5CE5-F0DF-40B7-8BA0-69DF395033C8",
              "collapseAllRules": [Function],
              "dispatcher": undefined,
              "expandAllRules": [Function],
              "source": undefined,
              "telemetryFactory": undefined,
              "toggleCardSelection": [Function],
              "toggleRuleExpandCollapse": [Function],
              "toggleVisualHelper": [Function],
            },
            "getPropertyConfigById": [Function],
          }
        }
        result={
          Object {
            "descriptors": Object {},
            "identifiers": Object {
              "conciseName": "test-concise-name",
              "identifier": "test-id",
              "this-property-does-not-have-config": "some value",
            },
            "isSelected": false,
            "resolution": Object {},
            "ruleId": "image-alt",
            "status": "fail",
            "uid": "some-guid-here",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`InstanceDetails renders 1`] = `
<div
  className="instanceDetailsCardContainer"
  data-automation-id="instance-card"
  role="table"
>
  <div
    aria-label="body img card"
    className="instanceDetailsCard"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="row"
    tabIndex={0}
  >
    <div
      role="gridcell"
    >
      <table
        className="reportInstanceTable"
      >
        <tbody>
          <React.Fragment>
            <css-selector
              deps={
                Object {
                  "cardInteractionSupport": Object {
                    "supportsCopyFailureDetails": true,
                    "supportsHighlighting": true,
                    "supportsIssueFiling": true,
                  },
                  "cardSelectionMessageCreator": proxy {
                    "___id": "BCDF5CE5-F0DF-40B7-8BA0-69DF395033C8",
                    "collapseAllRules": [Function],
                    "dispatcher": undefined,
                    "expandAllRules": [Function],
                    "source": undefined,
                    "telemetryFactory": undefined,
                    "toggleCardSelection": [Function],
                    "toggleRuleExpandCollapse": [Function],
                    "toggleVisualHelper": [Function],
                  },
                  "getPropertyConfigById": [Function],
                }
              }
              index={22}
              propertyData="body img"
            />
          </React.Fragment>
          <React.Fragment>
            <snippet
              deps={
                Object {
                  "cardInteractionSupport": Object {
                    "supportsCopyFailureDetails": true,
                    "supportsHighlighting": true,
                    "supportsIssueFiling": true,
                  },
                  "cardSelectionMessageCreator": proxy {
                    "___id": "BCDF5CE5-F0DF-40B7-8BA0-69DF395033C8",
                    "collapseAllRules": [Function],
                    "dispatcher": undefined,
                    "expandAllRules": [Function],
                    "source": undefined,
                    "telemetryFactory": undefined,
                    "toggleCardSelection": [Function],
                    "toggleRuleExpandCollapse": [Function],
                    "toggleVisualHelper": [Function],
                  },
                  "getPropertyConfigById": [Function],
                }
              }
              index={22}
              propertyData="this is a sample snippet"
            />
          </React.Fragment>
          <React.Fragment>
            <how-to-fix-web
              deps={
                Object {
                  "cardInteractionSupport": Object {
                    "supportsCopyFailureDetails": true,
                    "supportsHighlighting": true,
                    "supportsIssueFiling": true,
                  },
                  "cardSelectionMessageCreator": proxy {
                    "___id": "BCDF5CE5-F0DF-40B7-8BA0-69DF395033C8",
                    "collapseAllRules": [Function],
                    "dispatcher": undefined,
                    "expandAllRules": [Function],
                    "source": undefined,
                    "telemetryFactory": undefined,
                    "toggleCardSelection": [Function],
                    "toggleRuleExpandCollapse": [Function],
                    "toggleVisualHelper": [Function],
                  },
                  "getPropertyConfigById": [Function],
                }
              }
              index={22}
              propertyData={
                Object {
                  "all": Array [],
                  "any": Array [
                    "Element does not have an alt attribute",
                    "aria-label attribute does not exist or is empty",
                    "aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty",
                  ],
                  "none": Array [],
                }
              }
            />
          </React.Fragment>
        </tbody>
      </table>
      <InstanceDetailsFooter
        deps={
          Object {
            "cardInteractionSupport": Object {
              "supportsCopyFailureDetails": true,
              "supportsHighlighting": true,
              "supportsIssueFiling": true,
            },
            "cardSelectionMessageCreator": proxy {
              "___id": "BCDF5CE5-F0DF-40B7-8BA0-69DF395033C8",
              "collapseAllRules": [Function],
              "dispatcher": undefined,
              "expandAllRules": [Function],
              "source": undefined,
              "telemetryFactory": undefined,
              "toggleCardSelection": [Function],
              "toggleRuleExpandCollapse": [Function],
              "toggleVisualHelper": [Function],
            },
            "getPropertyConfigById": [Function],
          }
        }
        result={
          Object {
            "descriptors": Object {
              "snippet": "this is a sample snippet",
            },
            "identifiers": Object {
              "conciseName": "body img",
              "css-selector": "body img",
              "identifier": "body img",
            },
            "resolution": Object {
              "how-to-fix-web": Object {
                "all": Array [],
                "any": Array [
                  "Element does not have an alt attribute",
                  "aria-label attribute does not exist or is empty",
                  "aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty",
                ],
                "none": Array [],
              },
              "howToFixSummary": "sample how to fix summary",
            },
            "ruleId": "image-alt",
            "status": "fail",
            "uid": "some-guid-here",
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`InstanceDetails renders nothing when there is no card row configuration for the property / no property 1`] = `
<div
  className="instanceDetailsCardContainer"
  data-automation-id="instance-card"
  role="table"
>
  <div
    aria-label="test-id card"
    className="instanceDetailsCard"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="row"
    tabIndex={0}
  >
    <div
      role="gridcell"
    >
      <table
        className="reportInstanceTable"
      >
        <tbody>
          <React.Fragment />
          <React.Fragment />
          <React.Fragment />
        </tbody>
      </table>
      <InstanceDetailsFooter
        deps={
          Object {
            "cardInteractionSupport": Object {
              "supportsCopyFailureDetails": true,
              "supportsHighlighting": true,
              "supportsIssueFiling": true,
            },
            "cardSelectionMessageCreator": proxy {
              "___id": "BCDF5CE5-F0DF-40B7-8BA0-69DF395033C8",
              "collapseAllRules": [Function],
              "dispatcher": undefined,
              "expandAllRules": [Function],
              "source": undefined,
              "telemetryFactory": undefined,
              "toggleCardSelection": [Function],
              "toggleRuleExpandCollapse": [Function],
              "toggleVisualHelper": [Function],
            },
            "getPropertyConfigById": [Function],
          }
        }
        result={
          Object {
            "descriptors": Object {},
            "identifiers": Object {
              "conciseName": "test-concise-name",
              "identifier": "test-id",
              "this-property-does-not-have-config": "some value",
            },
            "resolution": Object {},
            "ruleId": "image-alt",
            "status": "fail",
            "uid": "some-guid-here",
          }
        }
      />
    </div>
  </div>
</div>
`;
